wrappers/services.go

```diff
@@ -471,13 +471,48 @@ func tryFileUpdate(path string, desiredContent []byte) (old *osutil.MemoryFileSt
 	}
 }
 
+type ServiceQuotaMap map[*snap.AppInfo]*quota.Group
+
+// MakeServiceQuotaMap builds a map of services and their quota groups. The goal is
+// to make sure that if any quota group applies to a snap service, then we can easily
+// look it up in this based on it's snap.AppInfo. We take the snap instance name, the services
+// want to include in the map (should only be services of the snap name provided) and the quota
+// group of the snap.
+func MakeServiceQuotaMap(snapName string, svcs []*snap.AppInfo, grp *quota.Group) ServiceQuotaMap {
+	if len(svcs) == 0 {
+		return nil
+	}
+
+	svcQuotaMap := make(ServiceQuotaMap, len(svcs))
+	for _, svc := range svcs {
+		// set nil grps for all services if parent is nil
+		if grp == nil {
+			svcQuotaMap[svc] = nil
+			continue
+		}
+
+		// always default to the snap quota group if the service is not
+		// in a separate one
+		svcGrp := grp.FindQuotaGroupForServiceInSubGroups(fmt.Sprintf("%s.%s", snapName, svc.Name))
+		if svcGrp == nil {
+			svcGrp = grp
+		}
+		svcQuotaMap[svc] = svcGrp
+	}
+	return svcQuotaMap
+}
+
 type SnapServiceOptions struct {
 	// VitalityRank is the rank of all services in the specified snap used by
 	// the OOM killer when OOM conditions are reached.
 	VitalityRank int
 
-	// QuotaGroup is the quota group for all services in the specified snap.
+	// QuotaGroup is the quota group for the specified snap.
 	QuotaGroup *quota.Group
```

Comment:
do we still use this btw?

Reply:
We don't and and we do. We have it for convenience as it is only used directly by `getExtraLayouts` in overlord/ifacestate/handlers.go to generate mount layouts for journal quota. The code *could* derive the quota group, but would have to manually parse the service map for it (so to avoid this we keep it).

But this actually brings us into an interesting challenge, because we only generate these mounts on a per-snap basis, so we cannot support journal namespaces for individual service groups. In other words services must inherit their journal namespace from the parent snap.

Right now we have not placed a limitation on this, but this needs to be implemented (but will add in a seperate PR to keep things lighter here).

Reply:
PR for limiting journal quota: https://github.com/snapcore/snapd/pull/12401